Improve README for August 2024 release (#196)
canonical/tdx: README.md

This tech preview of Intel TDX on Ubuntu 24.04 provides base host OS, guest OS, and remote attestation functionalities.
Follow these instructions to setup the Intel TDX host, create a TD, boot the TD, and attest the integrity of the TD's execution environment.

> The setup can be customized by editing the global configuration file: `setup-tdx-config`

Reviewer: Why do you want to remove this line ?

Author: I moved it to 4.2.2 which makes more sense in the sequence of setting up TDX.

Reviewer 2: I agree, but you should mention it again in the section 5.

Reviewer: I think we should keep it as it is before this MR, the config file will cover (now and in the future) all setup actions; i would like to have it at the beginning of the README

Author: How about a compromise?  Put it back like you requested, but also add it here with a little description as I proposed.  If it's at the beginning alone, I'm afraid it can be easily missed.

Reviewer: Ok

## 2. Report an Issue
Please submit issues [here](https://github.com/canonical/tdx/issues) and we'll get back to you ASAP.

> To help us with the debugging process, run the [system-report.sh](https://github.com/canonical/tdx/blob/noble-24.04/system-report.sh)

Reviewer: Requesting the user to run the system-report is already present when issue create an issue, i dont think it is necessary to repeat it here

Reviewer 2: I think it does not hurt to mention something like this multiple times. Many customers will still not do it.

Author: Agree with @BFuhry.

Reviewer: majority wins

tool and attach the report.  

<a id="supported-hardware"></a>
## 3. Supported Hardware
This release works with these Intel® Xeon® Processors:
* 4th Gen Intel® Xeon® Scalable Processors (select SKUs with Intel® TDX)
* 5th Gen Intel® Xeon® Scalable Processors

> * Intel® Xeon® 6 Processors with P-Cores

Reviewer: What is the Intel Xeon 6 Processors family ?

This tech preview of Intel TDX on Ubuntu 24.04 provides base host OS, guest OS, and remote attestation functionalities.
Follow these instructions to setup the Intel TDX host, create a TD, boot the TD, and attest the integrity of the TD's execution environment.

The host OS and TD setup can be cutomized by editing the global configuration file: `setup-tdx-config`.

<a id="report-an-issue"></a>
## 2. Report an Issue
Please submit issues [here](https://github.com/canonical/tdx/issues) and we'll get back to you ASAP.

> To help us with the debugging process, run the [system-report.sh](https://github.com/canonical/tdx/blob/noble-24.04/system-report.sh)

Author: ```suggestion
To help us with the debugging process, run the `system-report.sh`
```

> To help us with the debugging process, run the [system-report.sh](https://github.com/canonical/tdx/blob/noble-24.04/system-report.sh)

Author: fixed.

### 4.2 Enable Intel TDX in Host OS

1. Download this repository by downloading an asset file from the [releases page on GitHub](https://github.com/canonical/tdx/releases) or 
   by cloning the repository (at the appropriate tag/branch).
   
   For example: 
   ```bash

>    git clone https://github.com/canonical/tdx.git

Reviewer: A shorter version would be : `git clone -b noble-24.04 https://github.com/canonical/tdx.git`

   git checkout noble-24.04
   ```

> 2. Customize the setup by editing the global configuration file `setup-tdx-config`.  

Reviewer: This file `setup-tdx-config` is not only for the host setup customization, it contains the customizations for the guest image also;

Author: I changed it to "Customize the setup of the host and TD by editing the configuration..." to be more explicit.

   By default, remote attestation components are not installed on the host and inside the TD.  
   You can chose to install them automatically by setting `TDX_SETUP_ATTESTATION=1`.  

>    And if you chose to do that, you can skip [step 8.2.1](#step-8-2-1) and [step 8.3.3](#step-8-3-3).  

Author: changed.

2. Go to `Socket Configuration > Processor Configuration > TME, TME-MT, TDX`.

	* Set `Memory Encryption (TME)` to `Enable`
	* Set `Total Memory Encryption Bypass` to `Enable` (Optional setting for best host OS and regular VM performance.)
	* Set `Total Memory Encryption Multi-Tenant (TME-MT)` to `Enable`
	* Set `TME-MT memory integrity` to `Disable`
	* Set `Trust Domain Extension (TDX)` to `Enable`
	* Set `TDX Secure Arbitration Mode Loader (SEAM Loader)` to `Enable`. (NOTE: This allows loading Intel TDX Loader and Intel TDX Module from the ESP or BIOS.)
	* Set `TME-MT/TDX key split` to a non-zero value

3. Go to `Socket Configuration > Processor Configuration > Software Guard Extension (SGX)`.

> 	* Set `SW Guard Extensions (SGX)` to `Enable`

Author: I have seen "Enabled" on one machine, but all other machines is "Enable".  I think "Enable" is proper English.

> 	* Set `SW Guard Extensions (SGX)` to `Enable`

Reviewer 2: Scott and I checked a few other machines. On some Enabled/Disabled is used, on others Enable/Disable is used. So there is no right or wrong here. So your change is correct but also not necessary at the same time ;-)

sudo dmesg | grep -i tdx
```

> The message `virt/tdx: module initialized` proves TDX has properly initialized. Here is an example output:

Author: changed.

	./tdvirsh new
	```

   Details about `tdvirsh`:
   * To manage the lifecycle of TDs, we developed a wrapper around the `virsh` tool.
   This new `tdvirsh` tool extends `virsh` with new capabilities to create/remove TDs.
   * By default, `tdvirsh` will use an image located at `./image/tdx-guest-ubuntu-24.04-generic.qcow2` with the `generic` Ubuntu kernel.
   A different qcow2 image (e.g., one with an `intel` kernel) can be used by using the command-line option `-i IMAGE_PATH`.
   * By default, `tdvirsh` will use a XML libvirt template located at `./trust_domain.xml.template`.
   A different XML libvirt template can be used by using the command-line option `-t XML_PATH`.
   * All VMs can be listed with the following command:

	```bash
	./tdvirsh list --all
	```

	Example of output:

        Id   Name                                                        State
        ---------------------------------------------------------------------------

>         1    tdvirsh-trust_domain-f7210c2b-2657-4f30-adf3-639b573ea39f   running (ssh:32855, cid:3)

Author: not sure why, but this doesn't work.  i've already tested this.

>         1    tdvirsh-trust_domain-f7210c2b-2657-4f30-adf3-639b573ea39f   running (ssh:32855, cid:3)

Reviewer 2: What do you mean with it does not work? Currently, you do not have ```, so your version is missing formatting completely.

>         1    tdvirsh-trust_domain-f7210c2b-2657-4f30-adf3-639b573ea39f   running (ssh:32855, cid:3)

Author: very odd.  for me, with the ``` markers, it actually doesn't render properly.  it actually shows the ```.  that's why i removed it.  but i'll restore.

	crw-rw---- 1 root sgx     10, 124 Apr  3 21:14 /dev/sgx_vepc
	```

3. Verify the QGS service is running properly.
	```bash
	sudo systemctl status qgsd
	```

4. Verify the PCCS service is running properly.
	```bash
	sudo systemctl status pccs
	```

5. To setup the PCCS in the next step, you need a subscription to the [Intel PCS](https://api.portal.trustedservices.intel.com/provisioning-certification) service.
   1. [Subscribe](https://api.portal.trustedservices.intel.com/products#product=liv-intel-software-guard-extensions-provisioning-certification-service) to Intel PCS.
      For key rotation purpose, two subscription keys are generated.

>    2. [Retrieve](https://api.portal.trustedservices.intel.com/manage-subscriptions) one of your keys and use it in the next step.

Author: majority wins.  i find the original version too repetitive.  a lot of the same words, e.g. "both can be used for the following step."  Not worth haggling.

6. Configure the PCCS service.

	```bash
	sudo /usr/bin/pccs-configure
	```

	An example configuration:

	```
	Checking nodejs version ...
	nodejs is installed, continue...
	Checking cracklib-runtime ...
	Set HTTPS listening port [8081] (1024-65535) :
	Set the PCCS service to accept local connections only? [Y] (Y/N) :

> 	Set your Intel PCS API key (Press ENTER to skip) : b1b98494660fa0de93638c531d2c0df7

Reviewer: I dont think that we should put a valid API key here, each user will have his own key

Author: this is an invalid sample API key just for demo purpose.

	Choose caching fill method : [LAZY] (LAZY/OFFLINE/REQ) :
	Set PCCS server administrator password: <pccs-admin-password>
	Re-enter administrator password: <pccs-admin-password>
	Set PCCS server user password: <pccs-server-user-password>
	Re-enter user password: <pccs-server-user-password>
	Do you want to generate insecure HTTPS key and cert for PCCS service? [Y] (Y/N) :N
	```

   NOTE: The resulting config file is located at `/opt/intel/sgx-dcap-pccs/config/default.json`.

7. Restart the PCCS service.

	```bash
	sudo systemctl restart pccs
	```

8. Verify the PCCS service is running properly.

	```bash
	sudo systemctl status pccs
	```

9. Register the platform.

> 	The platform registration is done with the `mpa_registration_tool` tool.

Reviewer 2: ```suggestion
	Note that there are multiple alternatives to perform platform registration with different trade-offs and they are explained in detail in [Intel's Intel TDX Enabling Guide](https://cc-enabling.trustedservices.intel.com/intel-tdx-enabling-guide/02/infrastructure_setup/#platform-registration).
	In the following, we focus on the the direct registration variant that uses the Multi-package Registration Agent (MPA).
```

> 	This service is executed on system start up, registers the platform, and gets deactivated.

Reviewer 2: ```suggestion
	This agent is executed on system start up, registers the platform (if necessary), and gets deactivated.
```

	Please check the following two logs to confirm successful registration:

> 	1. Check the MPA service log.

Reviewer: It is rather : `Check the MPA service status`

Author: changed.

> 	1. Check the MPA service log.

Reviewer 2: ```suggestion
	1. Check MPA service log:
```
@hector-cao yes, this opens the service status, but the important piece is the log at the end of the status.

> 	1. Check the MPA service status.

Reviewer 2: ```suggestion
	1. Check the log of the MPA service:
```
It is important to mention "log" here, to properly refer to it later.

		```bash
		sudo systemctl status mpa_registration_tool
		```

> 		A successful registration example output:

Reviewer: I don't think we need the word `registration` here; the purpose of this command is only to verify the service is up and running

Author: Removed "registration"

> 		A successful registration example output:

Reviewer 2: ```suggestion
		Example output with successful registration in the log at the bottom:
```

	```bash
	sudo systemctl status pccs
	```

9. Platform registration.
9. Register the platform.

	The platform registration is done with the `mpa_registration_tool` tool.
	This service is executed on system start up, registers the platform, and gets deactivated.
	Please check the following two logs to confirm successful registration:

	1. Check service logs with following command:
	1. Check the MPA service log.

		```bash
		sudo systemctl status mpa_registration_tool
		```

		A successful example output:
		A successful registration example output:

> 		```bash

Reviewer 2: Undo change. This is still a result from bash

Author: i changed to ```console

		Apr 09 22:54:50 right-glider-515046 systemd[1]: Started mpa_registration_tool.service - Intel MPA Registratio>
		Apr 09 22:54:50 right-glider-515046 systemd[1]: mpa_registration_tool.service: Deactivated successfully.
		```

> 	2. Check the MPA log file.

Reviewer 2: ```suggestion
	2. Check the log file of the MPA:
```

		```bash
		cat /var/log/mpa_registration.log
		```

> 		A example output of a successful registration:

Reviewer 2: Example output with successful registration:

		A successful example output:
		A example output of a successful registration:

> 		```bash

Reviewer 2: Undo change. This is still a result from bash

Author: i'm changed to ```console

		[04-06-2024 03:05:54] INFO: Registration Flow - PLATFORM_ESTABLISHMENT or TCB_RECOVERY passed successfully.
		[04-06-2024 03:05:54] INFO: Finished Registration Agent Flow.
		```

> 	If an error is reported, re-do the registration from scratch with these steps:

Reviewer 2: ```suggestion
	If an error is reported in one of the logs, re-do the registration from scratch with these steps:
```
This is the place were the reference becomes clearer if you mention log above.

> 	If an error is reported, re-do with these steps:
> 	1. Clear the log with  `sudo rm /var/log/mpa_registration.log`.
> 	2. Reboot.
> 	3. Go into the BIOS.
> 	4. Navigate to `Socket Configuration > Processor Configuration > Software Guard Extension (SGX)`.
> 	5. Set these:
> 	  - `SGX Factory Reset` to `Enable`
> 	  - `SGX Auto MP Registration` to `Enable`

Reviewer 2: ```suggestion
    If an error is reported in one of the logs, you should begin registration from scratch with the following steps: 
    1. Remove the log file of the MPA:  `sudo rm /var/log/mpa_registration.log`.
    2. Reboot.
    3. Go into the BIOS.
    4. Navigate to `Socket Configuration > Processor Configuration > Software Guard Extension (SGX)`.
    5. Set these:
        - `SGX Factory Reset` to `Enabled`
        - `SGX Auto MP Registration` to `Enabled`
```

Author: took some of your suggestions.
